addon-test-support/index.js

```diff
@@ -0,0 +1,61 @@
+import { get } from '@ember/object';
+import { getContext, settled } from '@ember/test-helpers';
+import Promise from 'rsvp';
+import Test from 'ember-simple-auth/authenticators/test';
+
+const SESSION_SERVICE_KEY = 'service:session';
+const TEST_CONTAINER_KEY = 'authenticator:test';
+
+function ensureAuthenticator(owner) {
+  const authenticator = owner.lookup(TEST_CONTAINER_KEY);
+  if (!authenticator) {
+    owner.register(TEST_CONTAINER_KEY, Test);
+  }
+}
+
+/**
+ * Authenticates the session.
+ *
+ * @param {Object} sessionData Optional argument used to mock an authenticator
+ * response (e.g. a token or user).
+ * @return {Promise}
+ * @public
+ */
+export function authenticateSession(sessionData) {
+  const { owner } = getContext();
+  const session = owner.lookup(SESSION_SERVICE_KEY);
+  owner.setupRouter(); // router must initialize fully before authentication
```

Comment:
The prior versions of `authenticateSession` didn't do this, so I'm not 100% sure why this is needed.  Can you try to remove it and see what happens?

Reply:
I added `setupRouter` because visits (`visit()`) to routes protected by `AuthenticatedRouteMixin` fail without it.

TL;DR `hasRoute` is undefined when the authenticationSucceeded event is triggered by the `ApplicationRouteMixin`. I did a [full writeup of the series of calls](https://github.com/simplabs/ember-simple-auth/pull/1519#issuecomment-359837396) describing what occurs on the previous PR. @jordpo found a workaround was to invoke `visit` first to setup the router, but by calling `setupRouter` in the test-helper, it would initialize the routerMicroLib and fix the problem.

Reply:
Thank you for pointing me at that writeup!  Ultimately, this still seems like a bug being identified. IMHO, it is not logical for the `ApplicationRouteMixin` to call `.transitionTo` before the first navigation (e.g. before the first `visit()` call)...

/cc @marcoow @Turbo87

Reply:
> IMHO, it is not logical for the ApplicationRouteMixin to call .transitionTo before the first navigation (e.g. before the first visit() call)...

Yep, that's true. The problem is the [`ApplicationRouteMixin`'s `sessionAuthenticated` method](http://ember-simple-auth.com/api/classes/ApplicationRouteMixin.html#method_sessionAuthenticated) is called when the session is authenticated. Whether it makes sense to always do that automatically is certainly something that's debatable but that's how it is for now and we can't change that for now. We could maybe add a check into that method to just bail out early when the router is not ready (which I assume would require using private API to find out as well though 😞)

Reply:
@marcoow The `setupRouter` function already does that, so I opted not to add the additional complexity.

Reply:
@0xadada: what I meant is we can add a check for whether the router is ready [here](https://github.com/simplabs/ember-simple-auth/blob/1.4.2/addon/mixins/application-route-mixin.js#L108) and just return if it is not.

Reply:
Is this the last loose end?

Reply:
Yes, this is the last loose end. I guess we can leave it as is and see whether we can refactor the `ApplicationRouteMixin` separately so the error situation would never occur - that should be done in a separate PR anyways.

Reply:
Yes, totally agreed @marcoow. Calling `setupRouter` here should be fine for now. 

Let's open an issue to track refactoring the mixin so this isn't needed though...

Reply:
follow-up issue is here: https://github.com/simplabs/ember-simple-auth/issues/1539